draft-ietf-tls-tls13.md

```diff
@@ -2068,28 +2073,41 @@ selected_group
 : The group which the client MUST use for its new ClientHello.
 {:br }
 
-The "server_version", "cipher_suite" and "extensions" fields have the
-same meanings as their corresponding values in the ServerHello. The
-server SHOULD send only the extensions necessary for the client to
-generate a correct ClientHello pair.
-
-Upon receipt of a HelloRetryRequest, the client MUST first verify
-that the "selected_group" field does not identify a group which
-was not in the original ClientHello. If it was present, then
-the client MUST abort the handshake with a fatal "handshake_failure"
-alert. Clients SHOULD also abort with "handshake_failure" in response to any second
-HelloRetryRequest which was sent in the same connection (i.e.,
-where the ClientHello was itself in response to a HelloRetryRequest).
-
-Otherwise, the client MUST send a ClientHello with a new
-ClientKeyShare extension to the server. The ClientKeyShare MUST append
-a new ClientKeyShareOffer which is consistent with the
-"selected_group" field to the groups in the original ClientKeyShare.
-
-Upon re-sending the ClientHello and receiving the
-server's ServerHello/ServerKeyShare, the client MUST verify that
-the selected CipherSuite and NamedGroup match that supplied in
-the HelloRetryRequest.
+The server_version, cipher_suite, and extensions fields have the same
+meanings as their corresponding values in the ServerHello. The server
+SHOULD send only the extensions necessary for the client to generate a
+correct ClientHello pair. As with a ServerHello, only extensions first
+offered by the client can appear in a HelloRetryRequest.
+
+Upon receipt of a HelloRetryRequest, the client MUST first verify that
+the selected_group was offered in its "supported_groups" extension and
+that no KeyShare for the selected_group was offered in its "key_share"
+extension. If either of these conditions are detected, then the client
+MUST abort the handshake with a fatal "handshake_failure" alert.
+
+Clients SHOULD also abort with an "handshake_failure" alert in response
+to any second HelloRetryRequest which is sent on the same connection
+(i.e., where the ClientHello was itself in response to a
+HelloRetryRequest).
+
+If the HelloRetryRequest is verified as acceptable, the client MUST send
+a new ClientHello with an updated KeyShareOffer extension to the server.
+This KeyShareOffer MUST contain the exact previous set of offered keys
+with a new KeyShare value for the server's selected_group
+appended to the end of the list (or a new extension with exactly one
+value, if none was provided initially). All other fields in this new
+ClientHello MUST match the first attempt. Note that omitting any
+information previously included in the initial ClientHello introduces
+the risk of downgrade attack, as this retry exchange is unauthenticated.
```

Comment:
On Monday, October 12, 2015 07:22:04 pm ekr wrote:
> Please just merge the extensions in this PR, not add a bunch of other material.

This isn't directly part of this PR. It's PR #275, which this is based on top of. (all comments were addressed in that) If that's now on hold, I'll decouple this from it, do the renames in the existing text, and fix the merge conflicts later, then.